File: 백준/Gold/1707. 이분 그래프/이분 그래프.py
import sys

sys.setrecursionlimit(200000)

def dfs(graph, visited, cur):
  if visited[cur[0]] != -1:
    return visited[cur[0]] == cur[1]
    
  visited[cur[0]] = cur[1]
  
  for next in graph[cur[0]]:
    if visited[next] == visited[cur[0]]:
      return False
    if visited[next] == -1:
      if cur[1] == 0:
        if not dfs(graph, visited, (next, 1)):
          return False
      else:
        if not dfs(graph, visited, (next, 0)):
          return False
  return True
    

def solution():
  k = int(sys.stdin.readline())
  answer = []
  for _ in range(k):
    v, e = map(int, sys.stdin.readline().split())
    graph = [[] for _ in range(v + 1)]
    visited = [-1 for _ in range(v + 1)]
    for _ in range(e):
      a, b = map(int, sys.stdin.readline().split())
      graph[a].append(b)
      graph[b].append(a)
    result = True

    for idx in range(1, v + 1):
      if visited[idx] == -1:
        result = dfs(graph, visited, (idx, 0))
      if not result:
        break
    if result:
      answer.append("YES")
    else:
      answer.append("NO")
  for ans in answer:
    print(ans)
    
solution()

File: 백준/Gold/1707. 이분 그래프/이분 그래프.js
let fs = require('fs');
let input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');
const K = Number(input[0]);

let is2분;
let answer = '';

const dfs = (cur, color, graph) => {
    if (is2분 === 'NO') return;
    for (const next of graph[cur]) {
        if (color[next] !== '' && color[next] === color[cur]) {
            is2분 = 'NO'
            return;
        }
        if (color[next] === '') {
            color[next] = color[cur] === 'red' ? 'blue' : 'red';
            dfs(next, color, graph);
        }
    }
}

for (let i = 1; i < input.length; i++) {
    is2분 = 'YES';
    const [V, E] = input[i].split(' ').map(Number);
    const graph = Array.from({length: V + 1}, () => []);
    const color = Array.from({length: V + 1}, () => '');
    const limit = i + E;
    while (i < limit) {
        i++;
        const [x, y] = input[i].split(' ').map(Number);
        graph[x].push(y);
        graph[y].push(x);
    }
    for (let j = 1; j <= V; j++) {
        if(color[j] === '') {
            color[j] = 'red';
            dfs(j, color, graph);
        }
    }
    answer = answer + is2분 + '\n';
}

console.log(answer.trim());